### PR TITLE
Upgrade `k8s.io/*` to `v0.31`, `sigs.k8s.io/controller-runtime` to `v0.19`

### LINKED ISSUE #9667: Upgrade `k8s.io/*` to `v0.30`, `sigs.k8s.io/controller-runtime` to `v0.18`
**How to categorize this issue?**

/area open-source dev-productivity
/kind enhancement

**What would you like to be added**:

We should upgrade to the latest versions of our go upstream dependencies:
- `sigs.k8s.io/controller-runtime@v0.18.0`: https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.18.0
- `k8s.io/*@v0.30.0`

## Important changes / Action items
- [ ] admission.Decoder is now an interface (https://github.com/kubernetes-sigs/controller-runtime/pull/2736)
- [ ] Source, Event, Predicate, Handler: Add generics support (https://github.com/kubernetes-sigs/controller-runtime/pull/2783 https://github.com/kubernetes-sigs/controller-runtime/pull/2796)
- [ ] client: Fix SubResourceCreateOptions signature in subresource client (https://github.com/kubernetes-sigs/controller-runtime/pull/2766)
- [ ] Upgrade `sigs.k8s.io/controller-runtime/tools/setup-envtest` (ref https://github.com/gardener/gardener/pull/9685)
- [ ] Use new code-generator script (kube_codegen.sh). Currently the script has few issues https://github.com/kubernetes/kubernetes/issues/121247 due to which it cannot be used in our code-base.  

Here is a list of a few upstream changes to look out for when vendoring g/g, k/* and c-r in any of our repos (e.g. extensions). Please consider the release notes of [controller-runtime@v0.18](https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.18.0) and upwards as well for a complete list and more details.

### ⚠️ breaking / most notable changes 
- ...

### ℹ️ other changes / good to know / for information only:
- Text logging in Kubernetes components now uses textlogger https://github.com/kubernetes/kubernetes/pull/114672
- ...

### PR DESCRIPTION
**Pull request by ary1992**
**How to categorize this PR?**

/area open-source dev-productivity
/kind enhancement

**What this PR does / why we need it**:
Upgrade dependencies

- k8s.io/* to v0.31.0
- sigs.k8s.io/controller-runtime to v0.19.0

**Which issue(s) this PR fixes**:
Fixes #9667 

**Special notes for your reviewer**:
mockgen was failing where there was a need for generating mocks for the generics interfaces
```
> make generate WHAT="manifests" MANIFESTS_DIRS=third_party/mock/client-go/util/workqueue

For more info on the generate command, Run 'make generate PRINT_HELP=y'

> Generating manifests for folders: third_party/mock/client-go/util/workqueue
Generate github.com/gardener/gardener/third_party/mock/client-go/util/workqueue
# command-line-arguments
./prog.go:28:52: cannot use generic type pkg_.TypedRateLimitingInterface without instantiation
# command-line-arguments
./prog.go:28:52: cannot use generic type pkg_.TypedRateLimitingInterface without instantiation
prog.go:13:2: no required module provides package go.uber.org/mock/mockgen/model: go.mod file not found in current directory or any parent directory; see 'go help modules'
prog.go:15:2: no required module provides package k8s.io/client-go/util/workqueue: go.mod file not found in current directory or any parent directory; see 'go help modules'
2024/08/22 15:26:37 Loading input failed: exit status 1
third_party/mock/client-go/util/workqueue/doc.go:5: running "mockgen": exit status 1
make: *** [generate] Error 1
```

We are using `reflecting` mode in the mockgen to generate the mocks and it has an issue with the generics. Hence until the issue is fixed upstream, we will have to use `source` mode which requires files to be present locally. 

There is an PR open upstream to fix this.
- https://github.com/uber-go/mock/pull/198

Currently there are few issues hence in draft:

- [x] `k8s.io/kubernetes/cmd/import-boss` is pulling whole kubernetes/kubernetes under the go.mod. Because this vgopath was not working properly.
- [x] Make check command is failing
```
Executing golangci-lint
WARN [runner] Can't run linter goanalysis_metalinter: buildir: failed to load package fakeclient: could not load export data: no export data for "github.com/gardener/machine-controller-manager/pkg/fakeclient"
ERRO Running error: can't run linter goanalysis_metalinter
buildir: failed to load package fakeclient: could not load export data: no export data for "github.com/gardener/machine-controller-manager/pkg/fakeclient"
make: *** [check] Error 3
```
- [x] Make test command has error.
```
# github.com/gardener/machine-controller-manager/pkg/fakeclient
../../../../pkg/mod/github.com/gardener/machine-controller-manager@v0.53.1/pkg/fakeclient/client.go:209:34: cannot use t (variable of type *FakeObjectTracker) as "k8s.io/client-go/testing".ObjectTracker value in argument to w.dispatchInitialObjects: *FakeObjectTracker does not implement "k8s.io/client-go/testing".ObjectTracker (missing method Apply)
../../../../pkg/mod/github.com/gardener/machine-controller-manager@v0.53.1/pkg/fakeclient/client.go:447:57: cannot use o (variable of type *FakeObjectTracker) as "k8s.io/client-go/testing".ObjectTracker value in argument to k8stesting.ObjectReaction: *FakeObjectTracker does not implement "k8s.io/client-go/testing".ObjectTracker (missing method Apply)
../../../../pkg/mod/github.com/gardener/machine-controller-manager@v0.53.1/pkg/fakeclient/client.go:509:57: cannot use o (variable of type *FakeObjectTracker) as "k8s.io/client-go/testing".ObjectTracker value in argument to k8stesting.ObjectReaction: *FakeObjectTracker does not implement "k8s.io/client-go/testing".ObjectTracker (missing method Apply)
FAIL	github.com/gardener/gardener/cmd/machine-controller-manager-provider-local [build failed]
```

**Release note**:

```other developer
The following dependencies are updated:
- `k8s.io/*` : `v0.29.8` -> `v0.31.0`
- `sigs.k8s.io/controller-runtime`: `v0.17.5` -> `v0.19.0`
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Thanks for the PR @ary1992!
Do you already have a plan to resolve the open points mentioned in the PR description?

**Comment by ary1992** [comment]
> Thanks for the PR @ary1992! Do you already have a plan to resolve the open points mentioned in the PR description?

I am still looking into them. But the rest of the changes can be reviewed.

**Comment by MartinWeindel** [comment]
Suggested work around for the `import-boss` dependency issue: https://github.com/MartinWeindel/gardener/commit/6977e4bdba7f4b65e4daeec10b3c7b58186bf51b

**Comment by ary1992** [comment]
Kubernetes dependency upgrade PR on MCM https://github.com/gardener/machine-controller-manager/pull/938

File: example/provider-local/garden/base/patch-imagevector-overwrite.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+apiVersion: core.gardener.cloud/v1
```
Comment (rfranzke) [inline example/provider-local/garden/base/patch-imagevector-overwrite.yaml]: This file should not be committed to git.

File: go.mod [hunk 2]
```diff
@@ -58,49 +58,52 @@ require (
...
 )
 
+require dario.cat/mergo v1.0.1 // indirect
+
+require (
+	k8s.io/code-generator v0.31.1
+	k8s.io/kube-state-metrics/v2 v2.13.0
+	k8s.io/kubelet v0.29.5
+	k8s.io/metrics v0.29.5
+)
```
Comment (rfranzke) [inline go.mod]: Why is this in a separate `require` block? Can you merge this into above? Also, why do we request `v0.29.5` of `kubelet` and `metrics` instead of `v0.31`?

File: go.mod [hunk 3]
```diff
@@ -58,49 +58,52 @@ require (
...
-	sigs.k8s.io/controller-runtime v0.17.6
+	sigs.k8s.io/controller-runtime v0.19.0
 	sigs.k8s.io/controller-runtime/tools/setup-envtest v0.0.0-20231015215740-bf15e44028f9 // v0.16.3
-	sigs.k8s.io/controller-tools v0.14.0
+	sigs.k8s.io/controller-tools v0.16.1
 	sigs.k8s.io/structured-merge-diff/v4 v4.4.1
 	sigs.k8s.io/yaml v1.4.0
 )
 
+require dario.cat/mergo v1.0.1 // indirect
```
Comment (rfranzke) [inline go.mod]: same here

File: pkg/admissioncontroller/webhook/admission/authenticationconfig/handler.go [hunk 4]
```diff
@@ -237,3 +249,10 @@ func getAuthenticationConfiguration(cm *corev1.ConfigMap) (string, error) {
 	}
 	return authenticationConfigurationRaw, nil
 }
+
+func getDisallowedIssuers(kubeAPIServerConfig *gardencore.KubeAPIServerConfig) []string {
+	issuer := kubeAPIServerConfig.ServiceAccountConfig.Issuer
+	acceptedIssuers := kubeAPIServerConfig.ServiceAccountConfig.AcceptedIssuers
+
+	return append(acceptedIssuers, *issuer)
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhook/admission/authenticationconfig/handler.go]: This can be `nil`, no?

File: pkg/controller/networkpolicy/add.go [hunk 5]
```diff
@@ -65,34 +65,41 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, runt
 		r.additionalNamespaceLabelSelectors = append(r.additionalNamespaceLabelSelectors, selector)
 	}
 
+	predicates := []predicate.Predicate{
+		predicateutils.ForEventTypes(predicateutils.Create, predicateutils.Update),
+	}
```
Comment (rfranzke) [inline pkg/controller/networkpolicy/add.go]: Why did you move this here into a variable? Can you inline this again?

File: pkg/controller/vpaevictionrequirements/add.go [hunk 6]
```diff
@@ -42,19 +43,22 @@ func (r *Reconciler) AddToManager(mgr manager.Manager, seedCluster cluster.Clust
 		return fmt.Errorf("failed computing label selector predicate for eviction requirements managed by controller: %w", err)
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/controller/vpaevictionrequirements/add.go]: same here?

File: pkg/controllerutils/source_test.go [hunk 7]
```diff
@@ -22,35 +21,16 @@ import (
...
-			Expect(EnqueueOnce(ctx, nil, queue)).To(Succeed())
-			Expect(queue.Len()).To(Equal(1))
-
-			item, v := queue.Get()
-			Expect(item).To(Equal(reconcile.Request{}))
-			Expect(v).To(BeFalse())
-		})
-	})
-
-	Describe("#HandleOnce", func() {
```
Comment (rfranzke) [inline pkg/controllerutils/source_test.go:43]: Can you also add a small unit test for the replacement of `HandleOnce`?

File: pkg/gardenlet/controller/backupbucket/add.go [hunk 8]
```diff
@@ -49,6 +49,11 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, gard
 		r.GardenNamespace = v1beta1constants.GardenNamespace
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/backupbucket/add.go]: Same here

File: pkg/gardenlet/controller/backupentry/add.go [hunk 9]
```diff
@@ -52,6 +52,11 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, gard
 		r.GardenNamespace = v1beta1constants.GardenNamespace
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/backupentry/add.go]: Same here

File: pkg/gardenlet/controller/bastion/add.go [hunk 10]
```diff
@@ -43,6 +43,10 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, gard
 		r.Clock = clock.RealClock{}
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/bastion/add.go]: Same here

File: pkg/gardenlet/controller/controllerinstallation/care/add.go [hunk 11]
```diff
@@ -48,29 +48,35 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, gard
 		r.GardenNamespace = v1beta1constants.GardenNamespace
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/controllerinstallation/care/add.go]: Same here

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/add.go [hunk 12]
```diff
@@ -47,19 +47,22 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, gard
 		r.HelmRegistry = helmRegisty
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/controllerinstallation/controllerinstallation/add.go]: Same here

File: pkg/gardenlet/controller/gardenlet/add.go [hunk 13]
```diff
@@ -62,6 +63,11 @@ func (r *Reconciler) AddToManager(
 		r.ValuesHelper = gardenletdeployer.NewValuesHelper(&r.Config)
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/gardenlet/add.go]: Same here

File: pkg/gardenlet/controller/managedseed/add.go [hunk 14]
```diff
@@ -66,29 +66,33 @@ func (r *Reconciler) AddToManager(
 		r.GardenNamespaceShoot = v1beta1constants.GardenNamespace
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/managedseed/add.go]: Same here

File: pkg/gardenlet/controller/seed/care/add.go [hunk 15]
```diff
@@ -44,30 +44,35 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, gard
 		r.Clock = clock.RealClock{}
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/seed/care/add.go]: Same here

File: pkg/gardenlet/controller/seed/lease/add.go [hunk 16]
```diff
@@ -35,20 +38,23 @@ func (r *Reconciler) AddToManager(mgr manager.Manager, gardenCluster cluster.Clu
 		r.LeaseNamespace = gardencorev1beta1.GardenerSeedLeaseNamespace
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/seed/lease/add.go]: Same here

File: pkg/gardenlet/controller/seed/seed/add.go [hunk 17]
```diff
@@ -57,19 +58,22 @@ func (r *Reconciler) AddToManager(mgr manager.Manager, gardenCluster cluster.Clu
 		r.ClientCertificateExpirationTimestamp = &metav1.Time{Time: gardenletClientCertificate.Leaf.NotAfter}
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/gardenlet/controller/seed/seed/add.go]: Same here

File: pkg/gardenlet/controller/shoot/care/add.go [hunk 18]
```diff
@@ -38,18 +39,21 @@ func (r *Reconciler) AddToManager(mgr manager.Manager, gardenCluster cluster.Clu
 		r.Clock = clock.RealClock{}
 	}
 
+	predicates := []predicate.Predicate{r.ShootPredicate()}
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/care/add.go]: Same here

File: pkg/gardenlet/controller/shoot/state/add.go [hunk 19]
```diff
@@ -34,14 +35,17 @@ func (r *Reconciler) AddToManager(mgr manager.Manager, gardenCluster, seedCluste
 		r.Clock = clock.RealClock{}
 	}
 
+	predicates := []predicate.Predicate{r.SeedNameChangedPredicate()}
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/state/add.go]: Same here

File: pkg/nodeagent/controller/operatingsystemconfig/add.go [hunk 20]
```diff
@@ -53,16 +53,19 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager) erro
 		r.Extractor = registry.NewExtractor()
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/nodeagent/controller/operatingsystemconfig/add.go]: Same here

File: pkg/operator/controller/gardenlet/add.go [hunk 21]
```diff
@@ -64,21 +64,23 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, virt
 		}
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/operator/controller/gardenlet/add.go]: Same here

File: pkg/provider-local/controller/ingress/add.go [hunk 22]
```diff
@@ -41,13 +41,17 @@ func AddToManagerWithOptions(ctx context.Context, mgr manager.Manager, opts AddO
...
 	}
 
-	return ctrl.Watch(source.Kind(mgr.GetCache(), &networkingv1.Ingress{}), &handler.EnqueueRequestForObject{}, predicate.NewPredicateFuncs(func(object client.Object) bool {
-		namespace := &corev1.Namespace{}
-		if err := mgr.GetClient().Get(ctx, client.ObjectKey{Name: object.GetNamespace()}, namespace); err != nil {
-			ctrl.GetLogger().Error(err, "Unable to fetch namespace")
-		}
-		return namespace.Labels[v1beta1constants.GardenRole] == v1beta1constants.GardenRoleShoot
-	}))
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/provider-local/controller/ingress/add.go]: Same here

File: pkg/resourcemanager/controller/csrapprover/add.go [hunk 23]
```diff
@@ -31,21 +31,24 @@ func (r *Reconciler) AddToManager(mgr manager.Manager, sourceCluster, targetClus
 		r.TargetClient = targetCluster.GetClient()
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/resourcemanager/controller/csrapprover/add.go]: Same here

File: pkg/resourcemanager/controller/networkpolicy/add.go [hunk 24]
```diff
@@ -50,17 +50,20 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, targ
 		r.selectors = append(r.selectors, selector)
 	}
 
+	predicates := []predicate.Predicate{r.ServicePredicate()}
```
Comment (rfranzke) [inline pkg/resourcemanager/controller/networkpolicy/add.go]: Same here

File: pkg/resourcemanager/controller/node/agentreconciliationdelay/add.go [hunk 25]
```diff
@@ -25,14 +27,19 @@ func (r *Reconciler) AddToManager(mgr manager.Manager, targetCluster cluster.Clu
 		r.TargetClient = targetCluster.GetClient()
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/resourcemanager/controller/node/agentreconciliationdelay/add.go]: Same here

File: pkg/resourcemanager/controller/node/criticalcomponents/add.go [hunk 26]
```diff
@@ -32,16 +32,21 @@ func (r *Reconciler) AddToManager(mgr manager.Manager, targetCluster cluster.Clu
 		r.Recorder = targetCluster.GetEventRecorderFor(ControllerName + "-controller")
 	}
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/resourcemanager/controller/node/criticalcomponents/add.go]: Same here

File: pkg/resourcemanager/controller/tokeninvalidator/add.go [hunk 27]
```diff
@@ -42,6 +42,9 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, targ
 	secret := &metav1.PartialObjectMetadata{}
 	secret.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Secret"))
 
+	predicates := []predicate.Predicate{
```
Comment (rfranzke) [inline pkg/resourcemanager/controller/tokeninvalidator/add.go]: Same here

File: pkg/resourcemanager/webhook/extensionvalidation/add.go [hunk 28]
```diff
@@ -59,11 +59,11 @@ func AddToManager(mgr manager.Manager) error {
 		&extensionsv1alpha1.OperatingSystemConfig{}: &operatingSystemConfigValidator{},
 		&extensionsv1alpha1.Worker{}:                &workerValidator{},
 	} {
+		// RecoverPanic is defaulted to true is not set in the manager.
```
Comment (rfranzke) [inline pkg/resourcemanager/webhook/extensionvalidation/add.go]: The grammar seems weird.

**Comment by rfranzke** [comment]
/hold
until https://github.com/gardener/gardener/pull/10539 is merged

File: hack/update-codegen.sh [hunk 29]
```diff
@@ -307,22 +261,17 @@ export -f controllermanager_groups
...
-
-  bash "${CODE_GEN_DIR}"/generate-internal-groups.sh \
-    conversion \
-    github.com/gardener/gardener/pkg/client/admissioncontrollerconfig \
-    github.com/gardener/gardener/pkg/admissioncontroller/apis \
-    github.com/gardener/gardener/pkg/admissioncontroller/apis \
-    "config:v1alpha1" \
-    --extra-peer-dirs=github.com/gardener/gardener/pkg/admissioncontroller/apis/config,github.com/gardener/gardener/pkg/admissioncontroller/apis/config/v1alpha1,k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/apimachinery/pkg/conversion,k8s.io/apimachinery/pkg/runtime,k8s.io/component-base/config,k8s.io/component-base/config/v1alpha1 \
-    -h "${PROJECT_ROOT}/hack/LICENSE_BOILERPLATE.txt"
+  source "${CODE_GEN_DIR}/kube_codegen.sh"
```
Comment (shafeeqes) [inline hack/update-codegen.sh]: Why not source this only once at the top?

File: cmd/gardener-admission-controller/app/app.go [hunk 30]
```diff
@@ -96,6 +97,9 @@ func run(ctx context.Context, log logr.Logger, cfg *config.AdmissionControllerCo
 			Port:    cfg.Server.Webhooks.Port,
 			CertDir: cfg.Server.Webhooks.TLS.ServerCertDir,
 		}),
+		Controller: controllerconfig.Controller{
+			SkipNameValidation: ptr.To(true),
```
Comment (shafeeqes) [inline cmd/gardener-admission-controller/app/app.go]: Why do we need to skip this, We want every controller name to be unique, right? Can we not set it and use the default: false?

File: hack/update-protobuf.sh [hunk 31]
```diff
@@ -31,11 +31,27 @@ popd > /dev/null
...
 # core Google protobuf types
 go-to-protobuf \
+  --go-header-file=${PROJECT_ROOT}/hack/LICENSE_BOILERPLATE.txt \
+  --output-dir="${GOPATH}/src" \
+  --proto-import="${GOPATH}/src/k8s.io/kubernetes/staging/src" \
+  --proto-import="${GOPATH}/src/k8s.io/kubernetes/vendor" \
   --packages="$(IFS=, ; echo "${PACKAGES[*]}")" \
-  --apimachinery-packages='-k8s.io/apimachinery/pkg/util/intstr,-k8s.io/apimachinery/pkg/api/resource,-k8s.io/apimachinery/pkg/runtime/schema,-k8s.io/apimachinery/pkg/runtime,-k8s.io/apimachinery/pkg/apis/meta/v1,-k8s.io/apimachinery/pkg/apis/meta/v1beta1,-k8s.io/api/core/v1,-k8s.io/api/rbac/v1,-k8s.io/api/autoscaling/v1,-k8s.io/api/networking/v1,-k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1' \
-  --go-header-file=${PROJECT_ROOT}/hack/LICENSE_BOILERPLATE.txt
+  --apimachinery-packages=$(IFS=, ; echo "${APIMACHINERY_PKGS[*]}")
```
Comment (shafeeqes) [inline hack/update-protobuf.sh]: Nit: keep the empty line

File: third_party/mock/client-go/util/workqueue/default_rate_limiters_test.go [hunk 32]
```diff
@@ -0,0 +1,231 @@
+/*
```
Comment (shafeeqes) [inline third_party/mock/client-go/util/workqueue/default_rate_limiters_test.go]: We don't need to copy the test files

Comment (shafeeqes) [reply third_party/mock/client-go/util/workqueue/default_rate_limiters_test.go]: Also, we can mention why we are copying the files only for this Interface, ie; that generic type cannot be used without instantiation

Comment (shafeeqes) [reply third_party/mock/client-go/util/workqueue/default_rate_limiters_test.go]: And the generation works also with just the three files. `delaying_queue.go`, `queue.go` and `rate_limiting_queue.go`. You could remove the rest.

**Review by shafeeqes — COMMENTED** [review]
Thanks for the PR!

**Comment by shafeeqes** [comment]
> /hold until #10539 is merged

/unhold #10539 is merged

**Comment by rfranzke** [comment]
/hold

It seems this raises the minimum supported Kubernetes version for garden cluster. The new minimum version must be validated and documented.

**Comment by ary1992** [comment]
Looked into the failing test with @acumino .
With https://github.com/kubernetes/kubernetes/pull/123405 `ValidatingAdmissionPolicy` was graduated to GA. We are enabling the plugin [here](https://github.com/gardener/gardener/blob/321456d78627cef07ac206f84ae531c05c888952/plugin/pkg/plugins.go#L136) in our g/g repo. Hence in the cluster with k8s version < 1.30, gardener-apiserver is not able to come up as it not able to list `*v1.ValidatingAdmissionPolicy` as `admissionregistration.k8s.io/v1/validatingadmissionpolicies` API is not present in the older kubernetes version.
```
{"level":"info","ts":"2024-09-20T10:48:26.466Z","msg":"k8s.io/client-go@v0.31.1/tools/cache/reflector.go:243: failed to list *v1.ValidatingAdmissionPolicy: the server could not find the requested resource"}
{"level":"error","ts":"2024-09-20T10:48:26.466Z","msg":"Unhandled Error","logger":"UnhandledError","error":"k8s.io/client-go@v0.31.1/tools/cache/reflector.go:243: Failed to watch *v1.ValidatingAdmissionPolicy: failed to list *v1.ValidatingAdmissionPolicy: the server could not find the requested resource","stacktrace":"k8s.io/client-go/tools/cache.DefaultWatchErrorHandler\n\tk8s.io/client-go@v0.31.1/tools/cache/reflector.go:158\nk8s.io/client-go/tools/cache.(*Reflector).Run.func1\n\tk8s.io/client-go@v0.31.1/tools/cache/reflector.go:308\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\tk8s.io/apimachinery@v0.31.1/pkg/util/wait/backoff.go:226\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\tk8s.io/apimachinery@v0.31.1/pkg/util/wait/backoff.go:227\nk8s.io/client-go/tools/cache.(*Reflector).Run\n\tk8s.io/client-go@v0.31.1/tools/cache/reflector.go:306\nk8s.io/client-go/tools/cache.(*controller).Run.(*Group).StartWithChannel.func2\n\tk8s.io/apimachinery@v0.31.1/pkg/util/wait/wait.go:55\nk8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1\n\tk8s.io/apimachinery@v0.31.1/pkg/util/wait/wait.go:72"}
``` 

Hence we will have to disable the `ValidatingAdmissionPolicy` plugin if we want to use base clusters with k8s version < 1.30 or Update the base cluster k8s version to >= 1.30.

**Comment by rfranzke** [comment]
Fair enough @ary1992. Then let's simply not "enable this plugin by default" until our minimum supported garden version is 1.30 (you can comment this line and add a TODO statement).
Thanks for digging into it.

**Review by rfranzke — COMMENTED** [review]
/unhold
/approve

Let's go, thanks @ary1992 and everybody who contributed as well, well done!
